cpp/src/arrow/type.h

```diff
@@ -1367,6 +1367,12 @@ class ARROW_EXPORT DictionaryUnifier {
   /// after this is called
   virtual Status GetResult(std::shared_ptr<DataType>* out_type,
                            std::shared_ptr<Array>* out_dict) = 0;
+
+  /// \brief Return a result DictionaryType with the given index type.  If
+  /// the index type is not large enough then an invalid status will be returned.
+  /// The unifier cannot be used after this is called
+  virtual Status GetResultWithIndexType(std::shared_ptr<DataType> index_type,
```

Comment:
This could be an overload but the behavior was different enough I felt it warranted its own name.  `GetResult` is not actually used in the code base anywhere but `DictionaryUnifier` is an exported type.

Reply:
Correct, fixed.